File: iOS/EpicDash/EpicDash/Managers/BleManager.swift
import Foundation
import CoreBluetooth
import Combine

class BleManager: NSObject, ObservableObject {
    // MARK: - UUIDs (must match ESP32 firmware)
    static let serviceUUID = CBUUID(string: "4fafc201-1fb5-459e-8fcc-c5c9c331914b")
    static let buttonCharUUID = CBUUID(string: "beb5483e-36e1-4688-b7f5-ea07361b26a8")
    static let varDataCharUUID = CBUUID(string: "beb5483e-36e1-4688-b7f5-ea07361b26a9")
    static let varRequestCharUUID = CBUUID(string: "beb5483e-36e1-4688-b7f5-ea07361b26aa")
    static let gpsDataCharUUID = CBUUID(string: "beb5483e-36e1-4688-b7f5-ea07361b26ab")
    
    // GPS variable hashes for CAN transmission
    static let VAR_HASH_GPS_HMSD_PACKED: Int32 = 703958849
    static let VAR_HASH_GPS_MYQSAT_PACKED: Int32 = -1519914092
    static let VAR_HASH_GPS_ACCURACY: Int32 = -1489698215
    static let VAR_HASH_GPS_ALTITUDE: Int32 = -2100224086
    static let VAR_HASH_GPS_COURSE: Int32 = 1842893663
    static let VAR_HASH_GPS_LATITUDE: Int32 = 1524934922
    static let VAR_HASH_GPS_LONGITUDE: Int32 = -809214087
    static let VAR_HASH_GPS_SPEED: Int32 = -1486968225
    
    private static let deviceName = "ESP32 Dashboard"
    
    // MARK: - Published State
    @Published var isConnected = false
    @Published var isScanning = false
    @Published var logMessages: [String] = []
    
    // MARK: - Callbacks
    var onVariableData: ((Int, Float) -> Void)?
    
    // MARK: - Private
    private var centralManager: CBCentralManager!
    private var peripheral: CBPeripheral?
    private var buttonCharacteristic: CBCharacteristic?
    private var varDataCharacteristic: CBCharacteristic?
    private var varRequestCharacteristic: CBCharacteristic?
    private var gpsDataCharacteristic: CBCharacteristic?
    
    private var buttonWriteQueue: [Data] = []
    private var varRequestQueue: [Data] = []
    private var gpsDataQueue: [Data] = []
    private var isWritingButton = false
    private var isWritingVarRequest = false
    private var isWritingGpsData = false
    
    override init() {
        super.init()
        centralManager = CBCentralManager(delegate: self, queue: nil)
    }
    
    // MARK: - Public Methods
    func startScan() {
        guard centralManager.state == .poweredOn else {
            log("Bluetooth not ready")
            return
        }
        
        guard !isScanning else {
            log("Already scanning")
            return
        }
        
        isScanning = true
        log("Starting BLE scan...")
        
        centralManager.scanForPeripherals(
            withServices: [Self.serviceUUID],
            options: [CBCentralManagerScanOptionAllowDuplicatesKey: false]
        )
        
        // Stop scan after 10 seconds
        DispatchQueue.main.asyncAfter(deadline: .now() + 10) { [weak self] in
            self?.stopScan()
        }
    }
    
    func stopScan() {
        guard isScanning else { return }
        isScanning = false
        centralManager.stopScan()
        log("Scan stopped")
    }
    
    func disconnect() {
        if let peripheral = peripheral {
            centralManager.cancelPeripheralConnection(peripheral)
        }
        cleanup()
    }
    
    func sendButtonMask(_ mask: UInt16) {
        guard isConnected, buttonCharacteristic != nil else { return }
        
        var data = Data(count: 2)
        data[0] = UInt8(mask & 0xFF)
        data[1] = UInt8((mask >> 8) & 0xFF)
        
        buttonWriteQueue.append(data)
        processButtonQueue()
    }
    
    func requestVariable(_ varHash: Int) {
        guard isConnected, varRequestCharacteristic != nil else { return }
        
        var data = Data(count: 4)
        // Big-endian
        data[0] = UInt8((varHash >> 24) & 0xFF)
        data[1] = UInt8((varHash >> 16) & 0xFF)
        data[2] = UInt8((varHash >> 8) & 0xFF)
        data[3] = UInt8(varHash & 0xFF)
        
        varRequestQueue.append(data)
        processVarRequestQueue()
    }
    
    func requestVariablesBatch(_ varHashes: [Int]) {
        guard isConnected, varRequestCharacteristic != nil, !varHashes.isEmpty else { return }
        
        var data = Data(count: varHashes.count * 4)
        for (index, hash) in varHashes.enumerated() {
            let offset = index * 4
            data[offset] = UInt8((hash >> 24) & 0xFF)
            data[offset + 1] = UInt8((hash >> 16) & 0xFF)
            data[offset + 2] = UInt8((hash >> 8) & 0xFF)
            data[offset + 3] = UInt8(hash & 0xFF)
        }
        
        varRequestQueue.append(data)
        processVarRequestQueue()
    }
    
    // MARK: - GPS Data Methods
    func sendGpsData(varHash: Int32, value: Float) {
        guard isConnected, gpsDataCharacteristic != nil else { return }
        
        var data = Data(count: 8)
        // Hash (big-endian)
        data[0] = UInt8((varHash >> 24) & 0xFF)
        data[1] = UInt8((varHash >> 16) & 0xFF)
        data[2] = UInt8((varHash >> 8) & 0xFF)
        data[3] = UInt8(varHash & 0xFF)
        // Value (big-endian float)
        let valueBits = value.bitPattern.bigEndian
        withUnsafeBytes(of: valueBits) { bytes in
            data[4] = bytes[0]
            data[5] = bytes[1]
            data[6] = bytes[2]
            data[7] = bytes[3]
        }
        
        gpsDataQueue.append(data)
        processGpsDataQueue()
    }
    
    func sendGpsDataBatch(_ entries: [(hash: Int32, value: Float)]) {
        guard isConnected, gpsDataCharacteristic != nil, !entries.isEmpty else { return }
        
        var data = Data(count: entries.count * 8)
        for (index, entry) in entries.enumerated() {
            let offset = index * 8
            // Hash (big-endian)
            data[offset] = UInt8((entry.hash >> 24) & 0xFF)
            data[offset + 1] = UInt8((entry.hash >> 16) & 0xFF)
            data[offset + 2] = UInt8((entry.hash >> 8) & 0xFF)
            data[offset + 3] = UInt8(entry.hash & 0xFF)
            // Value (big-endian float)
            let valueBits = entry.value.bitPattern.bigEndian
            withUnsafeBytes(of: valueBits) { bytes in
                data[offset + 4] = bytes[0]
                data[offset + 5] = bytes[1]
                data[offset + 6] = bytes[2]
                data[offset + 7] = bytes[3]
            }
        }
        
        gpsDataQueue.append(data)
        processGpsDataQueue()
    }
    
    // MARK: - Private Methods
    private func cleanup() {
        peripheral = nil
        buttonCharacteristic = nil
        varDataCharacteristic = nil
        varRequestCharacteristic = nil
        gpsDataCharacteristic = nil
        buttonWriteQueue.removeAll()
        varRequestQueue.removeAll()
        gpsDataQueue.removeAll()
        isWritingButton = false
        isWritingVarRequest = false
        isWritingGpsData = false
        
        DispatchQueue.main.async {
            self.isConnected = false
        }
    }
    
    private func processButtonQueue() {
        guard !isWritingButton, !buttonWriteQueue.isEmpty,
              let char = buttonCharacteristic, let peripheral = peripheral else { return }
        
        let data = buttonWriteQueue.removeFirst()
        isWritingButton = true
        peripheral.writeValue(data, for: char, type: .withoutResponse)
        
        // For withoutResponse, manually trigger next
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.01) { [weak self] in
            self?.isWritingButton = false
            self?.processButtonQueue()
        }
    }
    
    private func processVarRequestQueue() {
        guard !isWritingVarRequest, !varRequestQueue.isEmpty,
              let char = varRequestCharacteristic, let peripheral = peripheral else { return }
        
        let data = varRequestQueue.removeFirst()
        isWritingVarRequest = true
        peripheral.writeValue(data, for: char, type: .withoutResponse)
        
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.01) { [weak self] in
            self?.isWritingVarRequest = false
            self?.processVarRequestQueue()
        }
    }
    
    private func processGpsDataQueue() {
        guard !isWritingGpsData, !gpsDataQueue.isEmpty,
              let char = gpsDataCharacteristic, let peripheral = peripheral else { return }
        
        let data = gpsDataQueue.removeFirst()
        isWritingGpsData = true
        peripheral.writeValue(data, for: char, type: .withoutResponse)
        
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.005) { [weak self] in
            self?.isWritingGpsData = false
            self?.processGpsDataQueue()
        }
    }
    
    private func log(_ message: String) {
        print("[BLE] \(message)")
        DispatchQueue.main.async {
            self.logMessages.append(message)
            if self.logMessages.count > 50 {
                self.logMessages.removeFirst()
            }
        }
    }
}

// MARK: - CBCentralManagerDelegate
extension BleManager: CBCentralManagerDelegate {
    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        switch central.state {
        case .poweredOn:
            log("Bluetooth powered on")
        case .poweredOff:
            log("Bluetooth powered off")
            cleanup()
        case .unauthorized:
            log("Bluetooth unauthorized")
        case .unsupported:
            log("Bluetooth unsupported")
        default:
            log("Bluetooth state: \(central.state.rawValue)")
        }
    }
    
    func centralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral, advertisementData: [String: Any], rssi RSSI: NSNumber) {
        let name = peripheral.name ?? "Unknown"
        
        if name == Self.deviceName || name.contains("ESP32") {
            log("Found device: \(name)")
            stopScan()
            
            self.peripheral = peripheral
            peripheral.delegate = self
            central.connect(peripheral, options: nil)
        }
    }
    
    func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        log("Connected to \(peripheral.name ?? "device")")
        peripheral.discoverServices([Self.serviceUUID])
    }
    
    func centralManager(_ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?) {
        log("Disconnected")
        cleanup()
    }
    
    func centralManager(_ central: CBCentralManager, didFailToConnect peripheral: CBPeripheral, error: Error?) {
        log("Failed to connect: \(error?.localizedDescription ?? "unknown")")
        cleanup()
    }
}

// MARK: - CBPeripheralDelegate
extension BleManager: CBPeripheralDelegate {
    func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        guard let services = peripheral.services else { return }
        
        for service in services {
            if service.uuid == Self.serviceUUID {
                log("Found service")
                peripheral.discoverCharacteristics([
                    Self.buttonCharUUID,
                    Self.varDataCharUUID,
                    Self.varRequestCharUUID,
                    Self.gpsDataCharUUID
                ], for: service)
            }
        }
    }
    
    func peripheral(_ peripheral: CBPeripheral, didDiscoverCharacteristicsFor service: CBService, error: Error?) {
        guard let characteristics = service.characteristics else { return }
        
        for char in characteristics {
            switch char.uuid {
            case Self.buttonCharUUID:
                buttonCharacteristic = char
                log("Found button characteristic")
            case Self.varDataCharUUID:
                varDataCharacteristic = char
                peripheral.setNotifyValue(true, for: char)
                log("Found variable data characteristic")
            case Self.varRequestCharUUID:
                varRequestCharacteristic = char
                log("Found variable request characteristic")
            case Self.gpsDataCharUUID:
                gpsDataCharacteristic = char
                log("Found GPS data characteristic")
            default:
                break
            }
        }
        
        if buttonCharacteristic != nil {
            log("Ready")
            DispatchQueue.main.async {
                self.isConnected = true
            }
        }
    }
    
    func peripheral(_ peripheral: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic, error: Error?) {
        guard characteristic.uuid == Self.varDataCharUUID,
              let data = characteristic.value else { return }
        
        // Parse batched response: multiple 8-byte entries [hash(4) + value(4)]
        var offset = 0
        while offset + 8 <= data.count {
            let hash = Int(Int32(bigEndian: data.subdata(in: offset..<offset+4).withUnsafeBytes { $0.load(as: Int32.self) }))
            let value = data.subdata(in: offset+4..<offset+8).withUnsafeBytes { $0.load(as: UInt32.self) }
            let floatValue = Float(bitPattern: UInt32(bigEndian: value))
            
            DispatchQueue.main.async {
                self.onVariableData?(hash, floatValue)
            }
            offset += 8
        }
    }
}

File: Android/app/src/main/java/com/buttonbox/ble/BleManager.kt
package com.buttonbox.ble

import android.annotation.SuppressLint
import android.bluetooth.*
import android.bluetooth.le.*
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.*

/**
 * BLE Manager for connecting to ESP32 Dashboard
 * Handles scanning, connecting, sending buttons, and receiving variable data
 */
class BleManager(private val context: Context) {

    companion object {
        private const val TAG = "BleManager"
        
        // Must match ESP32 firmware UUIDs
        val SERVICE_UUID: UUID = UUID.fromString("4fafc201-1fb5-459e-8fcc-c5c9c331914b")
        val CHAR_BUTTON_UUID: UUID = UUID.fromString("beb5483e-36e1-4688-b7f5-ea07361b26a8")
        val CHAR_VAR_DATA_UUID: UUID = UUID.fromString("beb5483e-36e1-4688-b7f5-ea07361b26a9")
        val CHAR_VAR_REQUEST_UUID: UUID = UUID.fromString("beb5483e-36e1-4688-b7f5-ea07361b26aa")
        val CHAR_GPS_DATA_UUID: UUID = UUID.fromString("beb5483e-36e1-4688-b7f5-ea07361b26ab")
        
        // GPS variable hashes for CAN transmission
        const val VAR_HASH_GPS_HMSD_PACKED = 703958849
        const val VAR_HASH_GPS_MYQSAT_PACKED = -1519914092
        const val VAR_HASH_GPS_ACCURACY = -1489698215
        const val VAR_HASH_GPS_ALTITUDE = -2100224086
        const val VAR_HASH_GPS_COURSE = 1842893663
        const val VAR_HASH_GPS_LATITUDE = 1524934922
        const val VAR_HASH_GPS_LONGITUDE = -809214087
        const val VAR_HASH_GPS_SPEED = -1486968225
        
        private const val DEVICE_NAME = "ESP32 Dashboard"
        private const val SCAN_TIMEOUT_MS = 10000L
    }

    interface BleCallback {
        fun onConnectionStateChanged(connected: Boolean)
        fun onLog(message: String)
        fun onScanResult(device: BluetoothDevice)
        fun onVariableData(varHash: Int, value: Float)
    }

    private var bluetoothAdapter: BluetoothAdapter? = null
    private var bluetoothGatt: BluetoothGatt? = null
    private var buttonCharacteristic: BluetoothGattCharacteristic? = null
    private var varDataCharacteristic: BluetoothGattCharacteristic? = null
    private var varRequestCharacteristic: BluetoothGattCharacteristic? = null
    private var gpsDataCharacteristic: BluetoothGattCharacteristic? = null
    private var callback: BleCallback? = null
    private var isScanning = false
    private val handler = Handler(Looper.getMainLooper())
    
    // Write queues for rapid operations
    private val buttonWriteQueue = ArrayDeque<ByteArray>()
    private val varRequestQueue = ArrayDeque<ByteArray>()
    private val gpsDataQueue = ArrayDeque<ByteArray>()
    private var isWritingButton = false
    private var isWritingVarRequest = false
    private var isWritingGpsData = false

    val isConnected: Boolean
        get() = bluetoothGatt != null && buttonCharacteristic != null

    fun setCallback(callback: BleCallback) {
        this.callback = callback
    }

    fun initialize(): Boolean {
        val bluetoothManager = context.getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager
        bluetoothAdapter = bluetoothManager?.adapter
        return bluetoothAdapter != null
    }

    @SuppressLint("MissingPermission")
    fun startScan() {
        if (isScanning) {
            log("Already scanning")
            return
        }

        val scanner = bluetoothAdapter?.bluetoothLeScanner
        if (scanner == null) {
            log("BLE scanner not available")
            return
        }

        isScanning = true
        log("Starting BLE scan...")

        handler.postDelayed({
            stopScan()
        }, SCAN_TIMEOUT_MS)

        val scanSettings = ScanSettings.Builder()
            .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
            .build()

        scanner.startScan(null, scanSettings, scanCallback)
    }

    @SuppressLint("MissingPermission")
    fun stopScan() {
        if (!isScanning) return
        isScanning = false
        bluetoothAdapter?.bluetoothLeScanner?.stopScan(scanCallback)
        log("Scan stopped")
    }

    private val scanCallback = object : ScanCallback() {
        @SuppressLint("MissingPermission")
        override fun onScanResult(callbackType: Int, result: ScanResult) {
            val device = result.device
            val name = device.name ?: return
            
            if (name == DEVICE_NAME || name.contains("ESP32")) {
                log("Found device: $name (${device.address})")
                stopScan()
                callback?.onScanResult(device)
                connect(device)
            }
        }

        override fun onScanFailed(errorCode: Int) {
            isScanning = false
            log("Scan failed: $errorCode")
        }
    }

    @SuppressLint("MissingPermission")
    fun connect(device: BluetoothDevice) {
        log("Connecting to ${device.name}...")
        bluetoothGatt = device.connectGatt(context, false, gattCallback, BluetoothDevice.TRANSPORT_LE)
    }

    @SuppressLint("MissingPermission")
    fun disconnect() {
        bluetoothGatt?.let { gatt ->
            log("Disconnecting...")
            gatt.disconnect()
            gatt.close()
        }
        bluetoothGatt = null
        buttonCharacteristic = null
        varDataCharacteristic = null
        varRequestCharacteristic = null
        gpsDataCharacteristic = null
        buttonWriteQueue.clear()
        varRequestQueue.clear()
        gpsDataQueue.clear()
        isWritingButton = false
        isWritingVarRequest = false
        isWritingGpsData = false
        callback?.onConnectionStateChanged(false)
    }

    private val gattCallback = object : BluetoothGattCallback() {
        @SuppressLint("MissingPermission")
        override fun onConnectionStateChange(gatt: BluetoothGatt, status: Int, newState: Int) {
            log("Connection state: newState=$newState status=$status")
            when (newState) {
                BluetoothProfile.STATE_CONNECTED -> {
                    log("Connected to GATT server")
                    // Small delay before service discovery for stability
                    handler.postDelayed({ gatt.discoverServices() }, 100)
                }
                BluetoothProfile.STATE_DISCONNECTED -> {
                    log("Disconnected from GATT server (status=$status)")
                    buttonCharacteristic = null
                    varDataCharacteristic = null
                    varRequestCharacteristic = null
                    gpsDataCharacteristic = null
                    handler.post { callback?.onConnectionStateChanged(false) }
                }
            }
        }

        @SuppressLint("MissingPermission")
        override fun onServicesDiscovered(gatt: BluetoothGatt, status: Int) {
            if (status != BluetoothGatt.GATT_SUCCESS) {
                log("Service discovery failed: $status")
                return
            }
            
            log("Services discovered")
            val service = gatt.getService(SERVICE_UUID)
            if (service == null) {
                log("Service not found")
                return
            }
            
            buttonCharacteristic = service.getCharacteristic(CHAR_BUTTON_UUID)
            varDataCharacteristic = service.getCharacteristic(CHAR_VAR_DATA_UUID)
            varRequestCharacteristic = service.getCharacteristic(CHAR_VAR_REQUEST_UUID)
            gpsDataCharacteristic = service.getCharacteristic(CHAR_GPS_DATA_UUID)
            
            // Log which characteristics were found
            log("Button char: ${if (buttonCharacteristic != null) "OK" else "MISSING"}")
            log("VarData char: ${if (varDataCharacteristic != null) "OK" else "MISSING"}")
            log("VarRequest char: ${if (varRequestCharacteristic != null) "OK" else "MISSING"}")
            log("GPS char: ${if (gpsDataCharacteristic != null) "OK" else "MISSING"}")
            
            // Enable notifications for variable data
            varDataCharacteristic?.let { char ->
                gatt.setCharacteristicNotification(char, true)
                val descriptor = char.getDescriptor(UUID.fromString("00002902-0000-1000-8000-00805f9b34fb"))
                descriptor?.let {
                    it.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
                    gatt.writeDescriptor(it)
                }
                log("Enabled variable data notifications")
            }
            
            if (buttonCharacteristic != null) {
                log("Found all characteristics")
                // Request high priority connection for faster throughput (7.5ms interval)
                gatt.requestConnectionPriority(BluetoothGatt.CONNECTION_PRIORITY_HIGH)
                // Request larger MTU for efficiency (default is 23, request 517)
                gatt.requestMtu(517)
            }
        }
        
        override fun onMtuChanged(gatt: BluetoothGatt, mtu: Int, status: Int) {
            log("MTU changed to $mtu (status: $status)")
            handler.post { callback?.onConnectionStateChanged(true) }
        }

        override fun onCharacteristicChanged(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic) {
            if (characteristic.uuid == CHAR_VAR_DATA_UUID) {
                val data = characteristic.value
                // Parse batched response: multiple 8-byte entries [hash(4) + value(4)]
                var offset = 0
                while (offset + 8 <= data.size) {
                    val hash = ByteBuffer.wrap(data, offset, 4).order(ByteOrder.BIG_ENDIAN).int
                    val value = ByteBuffer.wrap(data, offset + 4, 4).order(ByteOrder.BIG_ENDIAN).float
                    handler.post { callback?.onVariableData(hash, value) }
                    offset += 8
                }
            }
        }

        override fun onCharacteristicWrite(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic, status: Int) {
            when (characteristic.uuid) {
                CHAR_BUTTON_UUID -> {
                    isWritingButton = false
                    processButtonQueue()
                }
                CHAR_VAR_REQUEST_UUID -> {
                    isWritingVarRequest = false
                    processVarRequestQueue()
                }
                CHAR_GPS_DATA_UUID -> {
                    isWritingGpsData = false
                    processGpsDataQueue()
                }
            }
        }
    }

    @SuppressLint("MissingPermission")
    private fun processButtonQueue() {
        if (isWritingButton || buttonWriteQueue.isEmpty()) return
        
        val data = buttonWriteQueue.poll() ?: return
        val char = buttonCharacteristic ?: return
        val gatt = bluetoothGatt ?: return
        
        char.value = data
        char.writeType = BluetoothGattCharacteristic.WRITE_TYPE_NO_RESPONSE
        isWritingButton = gatt.writeCharacteristic(char)
    }

    @SuppressLint("MissingPermission")
    private fun processVarRequestQueue() {
        if (isWritingVarRequest || varRequestQueue.isEmpty()) return
        
        val data = varRequestQueue.poll() ?: return
        val char = varRequestCharacteristic ?: return
        val gatt = bluetoothGatt ?: return
        
        char.value = data
        char.writeType = BluetoothGattCharacteristic.WRITE_TYPE_NO_RESPONSE
        isWritingVarRequest = gatt.writeCharacteristic(char)
    }

    /**
     * Send button mask with queuing for rapid presses
     */
    fun sendButtonMask(buttonMask: Int): Boolean {
        if (!isConnected) return false
        
        val lowByte = (buttonMask and 0xFF).toByte()
        val highByte = ((buttonMask shr 8) and 0xFF).toByte()
        buttonWriteQueue.offer(byteArrayOf(lowByte, highByte))
        
        if (!isWritingButton) {
            processButtonQueue()
        }
        return true
    }

    /**
     * Request a variable value from ECU - queued for rapid requests
     */
    fun requestVariable(varHash: Int): Boolean {
        if (!isConnected) return false
        
        // Send hash as big-endian
        val data = ByteBuffer.allocate(4).order(ByteOrder.BIG_ENDIAN).putInt(varHash).array()
        varRequestQueue.offer(data)
        
        if (!isWritingVarRequest) {
            processVarRequestQueue()
        }
        return true
    }
    
    /**
     * Request multiple variables in one BLE write (batched for efficiency)
     * ESP32 will request each from CAN and batch the responses
     */
    fun requestVariablesBatch(varHashes: List<Int>): Boolean {
        if (!isConnected || varHashes.isEmpty()) return false
        
        // Pack all hashes into one buffer (4 bytes each, big-endian)
        val data = ByteBuffer.allocate(varHashes.size * 4).order(ByteOrder.BIG_ENDIAN)
        for (hash in varHashes) {
            data.putInt(hash)
        }
        varRequestQueue.offer(data.array())
        
        if (!isWritingVarRequest) {
            processVarRequestQueue()
        }
        return true
    }
    
    @SuppressLint("MissingPermission")
    private fun processGpsDataQueue() {
        if (isWritingGpsData || gpsDataQueue.isEmpty()) return
        
        val data = gpsDataQueue.poll() ?: return
        val char = gpsDataCharacteristic ?: return
        val gatt = bluetoothGatt ?: return
        
        char.value = data
        char.writeType = BluetoothGattCharacteristic.WRITE_TYPE_NO_RESPONSE
        isWritingGpsData = gatt.writeCharacteristic(char)
    }
    
    /**
     * Send GPS data to ESP32 for CAN transmission to ECU
     * Format: [0..3] VarHash (int32 BE), [4..7] Value (float32 BE)
     */
    fun sendGpsData(varHash: Int, value: Float): Boolean {
        if (!isConnected || gpsDataCharacteristic == null) return false
        
        val data = ByteBuffer.allocate(8).order(ByteOrder.BIG_ENDIAN)
            .putInt(varHash)
            .putFloat(value)
            .array()
        gpsDataQueue.offer(data)
        
        if (!isWritingGpsData) {
            processGpsDataQueue()
        }
        return true
    }
    
    /**
     * Send packed GPS data (uint32) to ESP32 for CAN transmission
     * Used for HMSD and MYQSAT packed values
     * The packed value is sent as raw bytes (not as float)
     */
    fun sendGpsDataPacked(varHash: Int, packedValue: Int): Boolean {
        if (!isConnected || gpsDataCharacteristic == null) return false
        
        val data = ByteBuffer.allocate(8).order(ByteOrder.BIG_ENDIAN)
            .putInt(varHash)
            .putInt(packedValue)
            .array()
        
        // Debug: print raw bytes
        val hexStr = data.joinToString(" ") { String.format("%02X", it) }
        log("GPS packed: hash=$varHash val=0x${Integer.toHexString(packedValue)} bytes=$hexStr")
        
        gpsDataQueue.offer(data)
        
        if (!isWritingGpsData) {
            processGpsDataQueue()
        }
        return true
    }
    
    /**
     * Send multiple GPS data entries in one BLE write (batched for efficiency)
     */
    fun sendGpsDataBatch(entries: List<Pair<Int, Float>>): Boolean {
        if (!isConnected) {
            log("GPS batch: not connected")
            return false
        }
        if (gpsDataCharacteristic == null) {
            log("GPS batch: no characteristic!")
            return false
        }
        if (entries.isEmpty()) return false
        
        val data = ByteBuffer.allocate(entries.size * 8).order(ByteOrder.BIG_ENDIAN)
        for ((hash, value) in entries) {
            data.putInt(hash)
            data.putFloat(value)
            log("GPS queue: hash=$hash val=$value")
        }
        
        // Debug: print raw bytes
        val bytes = data.array()
        val hexStr = bytes.joinToString(" ") { String.format("%02X", it) }
        log("GPS raw bytes: $hexStr")
        
        gpsDataQueue.offer(bytes)
        
        if (!isWritingGpsData) {
            processGpsDataQueue()
        }
        return true
    }

    private fun log(message: String) {
        Log.d(TAG, message)
        handler.post { callback?.onLog(message) }
    }
}
